utils/bitfield/src/lib.rs

```diff
@@ -1,462 +1,247 @@
-            if b {
-                callback(i)?;
-            }
-        }
-        Ok(())
+    /// Adds the bit at a given index to the bit field.
+    pub fn set(&mut self, bit: usize) {
+        self.unset.remove(&bit);
+        self.set.insert(bit);
     }
 
-    /// Returns true if there are no bits set, false if the bitfield is empty.
-    pub fn is_empty(&mut self) -> Result<bool> {
-        for b in self.as_mut_flushed()?.iter() {
-            if b == &true {
-                return Ok(false);
-            }
-        }
-
-        Ok(true)
+    /// Removes the bit at a given index from the bit field.
+    pub fn unset(&mut self, bit: usize) {
+        self.set.remove(&bit);
+        self.unset.insert(bit);
     }
 
-    /// Returns a slice of the bitfield with the start index of set bits
-    /// and number of bits to include in slice.
-    pub fn slice(&mut self, start: u64, count: u64) -> Result<BitField> {
-        if count == 0 {
-            return Ok(BitField::default());
+    /// Returns `true` if the bit field contains the bit at a given index.
+    pub fn get(&self, index: usize) -> bool {
+        if self.set.contains(&index) {
+            true
+        } else if self.unset.contains(&index) {
+            false
+        } else {
+            self.bitvec.get(index)
         }
-
-        // These conversions aren't ideal, but we aren't supporting 32 bit targets
-        let mut start = start as usize;
-        let mut count = count as usize;
-
-        let bitvec = self.as_mut_flushed()?;
-        let mut start_idx: usize = 0;
-        let mut range: usize = 0;
-        if start != 0 {
-            for (i, v) in bitvec.iter().enumerate() {
-                if v == &true {
-                    start -= 1;
-                    if start == 0 {
-                        start_idx = i + 1;
-                        break;
-                    }
-                }
-            }
-        }
-
-        for (i, v) in bitvec[start_idx..].iter().enumerate() {
-            if v == &true {
-                count -= 1;
-                if count == 0 {
-                    range = i + 1;
-                    break;
-                }
-            }
-        }
-
-        if count > 0 {
-            return Err("Not enough bits to index the slice");
-        }
-
-        let mut slice = BitVec::with_capacity(start_idx + range);
-        slice.resize(start_idx, false);
-        slice.extend_from_slice(&bitvec[start_idx..start_idx + range]);
-        Ok(BitField::Decoded(slice))
     }
 
-    /// Retrieves number of set bits in the bitfield
-    ///
-    /// This function requires a mutable reference for now to be able to handle the cached
-    /// changes in the case of an RLE encoded bitfield.
-    pub fn count(&mut self) -> Result<usize> {
-        Ok(self.as_mut_flushed()?.count_ones())
+    /// Returns the index of the lowest bit present in the bit field.
+    pub fn first(&self) -> Option<usize> {
+        self.iter().next()
```

Comment:
`flatten` is lazy, so I think both versions will compile the exact same way. There might still be a better way to implement this (and `bits` in general) though — `ranges` has to fully go through the `set` and `unset` bits up front, but `bits` could get away with only going through `set` up front and check `unset` containment lazily... I have to think about this more

Reply:
yeah fine to keep this way, I was just more double checking, seemed like there was a bit of overlapping logic because `ranges()` handles the set and unset bits, but iter also merges with the set bits. I would think the compiler would handle ignoring this well, but I'm wondering why `iter` merges with set bits and filters unset bits if handled already